### PR TITLE
Enforce Archive Limits

### PR DESCRIPTION
**Pull request by mjhilton**
Place reasonable limits on archive decompression so that we don't fill up the entire disk of a worker or deployment target

### CODE REVIEW

File: source/Calamari.Common/Features/Packages/Decorators/PackageEntryExtractorDecorator.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+namespace Calamari.Common.Features.Packages.Decorators
+{
+    /// <summary>
+    /// A base Decorator which allows addition to the behaviour of an IPackageEntryExtractor 
+    /// </summary>
+    /// <remarks>
+    /// IPackageEntryExtractor is a more specific IPackageExtractor, which provides a hook to the extraction
+    /// of each individual ArchiveEntry.
+    /// </remarks>
+    public class PackageEntryExtractorDecorator : IPackageEntryExtractor
```
Comment (mjhilton) [inline source/Calamari.Common/Features/Packages/Decorators/PackageEntryExtractorDecorator.cs:12]: Whoops, this is currently unused as I removed the only decorator that was deriving from it. The functionality and abstractions needed a rethink.

File: source/Calamari.Common/Features/Packages/Decorators/ArchiveLimits/EnforceCompressionRatioDecorator.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+                throw archiveLimitException!;
+
+            return base.Extract(packageFile, directory);
+        }
+
+        bool TryVerifyCompressionRatio(string packageFile, out ArchiveLimitException? exception)
+        {
+            try
+            {
+                if (maximumCompressionRatio >= 1)
```
Comment (mik-ky) [inline source/Calamari.Common/Features/Packages/Decorators/ArchiveLimits/EnforceCompressionRatioDecorator.cs:31]: I'm assuming this is more of a safety belt to ensure that this maximum compression ratio is set (as intended)? 👍 👍

**Review by mik-ky — APPROVED** [review]
LGTM  👍 

I  🐰 🕳️  reading through zip bombs while reviewing this.

**Comment by mjhilton** [comment]
Rebased on master, and modified the tests to cover the feature flags.

File: source/Calamari.Common/Features/Packages/Decorators/PackageEntryExtractorDecorator.cs (hunk 1 above)
Comment (mjhilton) [reply source/Calamari.Common/Features/Packages/Decorators/PackageEntryExtractorDecorator.cs:12]: This needs to be removed, but I'll do in a follow up PR.

File: source/Calamari.Common/Features/Packages/Decorators/ArchiveLimits/ArchiveLimitException.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
+﻿using System;
+using Calamari.Common.Plumbing.Extensions;
+
+namespace Calamari.Common.Features.Packages.Decorators.ArchiveLimits
+{
+    public class ArchiveLimitException : InvalidOperationException
+    {
+        public const string CorruptedHeaderWarning = "The given archive may be corrupt, or a security risk: the archive headers did not match the actual file sizes. As a precaution, those files have been truncated to their declared size. Please see https://oc.to/ArchiveDecompressionLimits for information on why we do this, and how to modify these settings if the archive is legitimate.";
```
[the remark this answers is not in the document]
Comment (mjhilton) [reply source/Calamari.Common/Features/Packages/Decorators/ArchiveLimits/ArchiveLimitException.cs:8]: We can create and edit that shortlink separately to the underlying thing it points to.

File: source/Calamari.Common/Features/Packages/Decorators/ArchiveLimits/EnforceCompressionRatioDecorator.cs (hunk 2 above)
Comment (mjhilton) [reply source/Calamari.Common/Features/Packages/Decorators/ArchiveLimits/EnforceCompressionRatioDecorator.cs:31]: Yep just ensuring nothing nonsense is provided.

**Pull request merged**